tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,391 @@
+
+    return updated_mask
+
+  def _grow_connections(self, step, weight, mask, grow_scores, dropped_mask, n_update, n_total):
+    """Following the dropping of connections, grow according to the grow scores.
+
+    Args:
+      step: current optimizer step
+      weight: current state of weights being optimized
+      mask: current mask pre-updates
+      grow_scores: gradient scores (via pruner specific metrics)
+      dropped_mask: mask reflecting dropped connections
+      n_update: number of connections to grow (based on drop ratio)
+      n_total: current total number of unmasked connections available
+    Returns:
+      grow_tensor: mask of values to set for new connections that were regrown
+      grown_mask: mask of grown connections (should contain n_update many 1s)
+      new_connections: grown connections based on previously dropped this iteration
+    """
+    # flatten the scores
+    grow_scores = tf.reshape(grow_scores, (-1,))
+    # set enabled connections (ones) to min(scores) - 1, i.e. they have the lowest scores
+    grow_scores_lifted = tf.where(
+      tf.math.equal(dropped_mask, 1),
+      tf.ones_like(dropped_mask) * (tf.reduce_min(grow_scores) - 1), grow_scores
+    )
+    grown_mask = self._generic_top_k(grow_scores_lifted, mask, n_update, n_total)
+    # ensure that masks are disjoint
+    tf.debugging.Assert(
+      tf.math.equal(tf.reduce_sum(dropped_mask * grown_mask), 0.), [dropped_mask, grown_mask])
+
+    grown_mask_reshaped = tf.reshape(grown_mask, mask.shape)
+    # set the values of the new connections
+    grow_tensor = self._get_grow_tensor(weight, self._grow_init_method, step=step)
+    new_connections = self._get_new_connections(self._drop_regrow_reinit, grown_mask_reshaped, mask)
+
+    return grow_tensor, grown_mask, new_connections
+
+
+  def _update_mask(self, step, update_fraction, mask, weight, grad):
+    """Called by _maybe_update_block_mask.
+    Updates mask based on weight and grad information.
+
+    Args:
+      step: current iteration from optimizer
+      update_fraction: the fraction of existing connections to update
+      mask: the mask to update
+      weight: trainable variable representing weights
+      grad: gradients obtained from the optimizer
+    """
+    # compute the top k magnitudes then update the current mask
+    drop_scores = self._get_drop_weights(mask, weight, noise_std=self._noise_std, step=step)
+    # need access to exactly which entries are growing to zero out optimizer slot
+    grow_scores = self._get_grow_grads(mask, grad)
+    n_total = tf.size(drop_scores)
+    n_ones = tf.cast(tf.reduce_sum(mask), dtype=tf.int32) # floor not ceiling like sparsity
+    n_prune = tf.cast(
+      tf.cast(n_ones, dtype=tf.float32) * update_fraction, tf.int32
+    )
+    n_keep = n_ones - n_prune
+
+    dropped_mask = self._generic_top_k(drop_scores, mask, n_keep, n_total)
+
+    if grow_scores is not None: # case where there is no growing
```

Comment:
I've unresolved the resolved conversation about this: 
```
tomerk 10 hours ago Owner
How would grow_scores ever be None if it's defined as math.abs(grads)?

Also, this block of code here (to build a grow mask) is what should be put in a util method, not just math.abs(grads).

 @xwinxu
xwinxu 4 hours ago • 
 Author Collaborator
it would be None if we didn't want to grow (i.e. if we jsut want to do pruning and no growing - which would be another method that's a variant of rigl) -- this is would be specified in another flag though that I'll add in the future.

 @xwinxu
xwinxu 4 hours ago Author Collaborator
so i.e. the variant would subclass this rigl pruner and override the get_grow_grads function, which could be to return None if it doesn't wish to grow.
```
It's a little pre-emptive to make an API decision like this, especially given there's lots of ways customizable grow scores could be implemented and there are no such variants being implemented or tested currently. (e.g. a `grow_score_fn` could be passed to the constructor, a boolean flag `grow_enabled` could be used, etc.)

I'd prefer inlining the grow_scores fn for now & removing this `grow_scores is not None` check, but feel free to put in a TODO requesting we follow up on customizing the `grow_scores_fn` in the future.